### ISSUE TITLE
Blender 2.80 support

### ISSUE THREAD

**Issue by MrApplejuice** [OGRECave/blender2ogre#33]
for reference: https://wiki.blender.org/wiki/Reference/Release_Notes/2.80/Python_API

Hello everybody, I wanted to use the plugin with the new blender 2.80, but - hey - as you would expect, it did not work.

So I put in my elbow-grease to patch the current version of the exporter. If somebody wants to have a look, check out this branch: `2.80-support`

I know that basic mesh and material exporting works again for 2.80 after this patch, but the following features remain untested for now:

- Textures
    - [ ] Texture image conversion (internal vs external images)
    - [x] Texture UV maps in general
    - [x] Multi UV map support
- [ ] Everything physics related (I actually know that I removed some stuff there)
- [ ] Custom variables on objects
- [ ] Scene support (I was only interested in exporting single models)
- [ ] I did not test whether the examples in the `/examples` folder still work (probably they do not)

But it is a starting point in case somebody wants to continue the conversion. Keep in mind, that I was not too keen on keeping backwards compatibility with older blender versions. I think that it will be hard to do so because the internal blender models for representing meshes have changed considerably and the code is currently not written in a way to include different exporters for different blender versions in an easy way.

**Comment by paroj** [comment]
thanks for the initial work. can you create a proper pull request out of this?

**Comment by MrApplejuice** [comment]
I am currently working on getting a somewhat decent implementation of the texture export feature in here - I feel that is pretty much a bare-minimum for this to be useful for Ogre-users. I am happy to make a proper pull-request, but I would like to emphasize that the current version will:
- Break any backwards compatibility with older blender versions
- The current version is quite buggy even with blender 2.80

Right now, it is quite late for me already. I will try to find some time over the weekend to make create a proper PR

**Comment by kromar** [comment]
@MrApplejuice does animation export work on this branch?

**Comment by MrApplejuice** [comment]
Hey there! I barely managed to get mesh exporting to work. I have tested nothing else. Not even the export of texture maps works right now (only the uv maps work). There were quite a lot of changes between blender 2.7x and 2.8x so I do not expect it to work. You can try tough. Maybe it is easy to fix. Don't know.

**Comment by Tectu** [comment]
I have tried different versions of Blender and blender2ogre. Blender 2.8 together with the corresponding `2.80-support` branch of this tool worked best so far.
It seems that for most very basic workflows this might be the best solution so far - except for the missing support for textures.

I never wrote an add-on for Blender so it's difficult for me to estimate how much work & what the complexity of said work would be.
Could anybody elaborate on this a bit? What is the current priority of adding texture export support to the 2.8 compatible version of this very essential tool?

**Comment by paroj** [comment]
in 2.8x "Blender Render" Materials, which closely resembled Ogre materials were replaced by nodes. Among other changes, one now has to search these nodes for something that looks like a texture. Currently nodes are not handled at all.

**Comment by Tectu** [comment]
> Thinking out of the box, it might be possible to analyze the node graphs and write a node-to-shader-program renderer.

I might be able to at least help with that. I'm the author of [QSchematic](https://github.com/simulton/QSchematic) so I'm somewhat used to working with graphs - especially analyzing them. But I guess we'd need more man power especially on the side of extending the actual Blender add-on code.

**Comment by paroj** [comment]
I had a thought that other exporters are faced with the same problem and, indeed, blender provides an easy wrapper for our usecase:
https://github.com/blender/blender-addons/blob/084eec36977c9a084aa020031f4a1aa0f0352170/io_scene_obj/export_obj.py#L70

**Comment by Tectu** [comment]
So, will we be blessed with a working blender2ogre plugin that supports Blender 2.80 including texture exporting soon? :p

**Comment by paroj** [comment]
> So, will we be blessed with a working blender2ogre plugin that supports Blender 2.80 including texture exporting soon? :p

whats your schedule? I am fine with using blender 2.79 for now

**Comment by MrApplejuice** [comment]
> I had a thought that other exporters are faced with the same problem and, indeed, blender provides an easy wrapper for our usecase:
> https://github.com/blender/blender-addons/blob/084eec36977c9a084aa020031f4a1aa0f0352170/io_scene_obj/export_obj.py#L70

Wow! That is a great find, indeed. When I find time to resume my work on my 3D Catan client, I can have a look, but that might take a while. Anyway great news :+1:

**Comment by Grodou** [comment]
Hello, I managed to export a simple image texture inside a material file, with great help from the Wavefront export script mentioned by @MrApplejuice (io_scene_obj/export_obj.py).
I don't really know OGRE formats, but if the only job is to adapt material.py, maybe I can help.
For a personal use, I added a UI checkbox to make texture file path relative to the export folder, rather than to the .blend file.
Don't know if it's a good idea...
